src/main/java/com/google/devtools/build/lib/rules/cpp/ClangHeaderMap.java

```diff
@@ -0,0 +1,325 @@
+package com.google.devtools.build.lib.rules.cpp;
+
+import java.io.File;
+import java.io.FileOutputStream;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.List;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.channels.FileChannel;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static java.lang.Math.max;
+
+public final class ClangHeaderMap {
+  // Logical representation of a bucket.
+  // The actual data is stored in the string pool.
+  private class HMapBucket {
```

Comment:
Nit: Let's make this comment a javadoc.

Reply:
And consider renaming to HeaderMapBucket?

Reply:
Sure, I'll rename it to HeaderMapBucket